---
title: "R Notebook"
output: html_notebook
---

Pilot data Parse bio 4 iPSC lines NPCs, 3 batches

Load libraries
```{r}
library(Seurat)
library(tidyverse)
#library(CelltypeR)

```


```{r}

# read in the data
# output from Parse bio pipeline adapted by Saeid
# run by Taylor
# data object made by Taylor
seu <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/ParseExample/Experiment1-mini12/Seurat_Parse12sample.RDS")

# this data object has not been filtered
# it does have PCA and variable features
seu

```
```{r}

seu[["percent.mt"]] <- PercentageFeatureSet(seu, pattern = "^MT-")
Idents(seu) <- 'orig.ident'
plot <- VlnPlot(seu, pt.size = 0.001, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
plot

Idents(seu) <- 'sample'
VlnPlot(seu, pt.size = 0.001, features = c("nFeature_RNA"))
VlnPlot(seu, pt.size = 0.001, features = "nCount_RNA")
VlnPlot(seu, pt.size = 0.001, features = "percent.mt")



```

Add in labels for batch and disease status and line

```{r}
#library("CelltypeR")
# CelltypeR library is a library I (Rhalena) made for flow cytometry but uses the seurat object and I made a quick add annotations function.

Seurat_Parse12sample <- seu

# here is the function
annotate <- function(seu, annotations, to_label, annotation_name = "CellType"){
  Idents(seu) <- to_label
  names(annotations) <- levels(seu)
  seu <- RenameIdents(seu, annotations)
  seu <- AddMetaData(object=seu, metadata=Idents(seu), col.name = annotation_name)

}

Idents(Seurat_Parse12sample) <- "sample"
sample.levels <- levels(Seurat_Parse12sample)
# should give the order of sample

# test
Seurat_Parse12sample <- annotate(Seurat_Parse12sample, annotations = sample.levels, to_label = "sample",annotation_name = "sample.test")
table(Seurat_Parse12sample$sample.test)
table(Seurat_Parse12sample$sample)
# these match

#input vector we got from the seurat object
# Define regular expression to match first part of the string
pattern <- "^[A-Za-z]+"

# Use gsub() to replace the first part of the string with an empty string
sample.levels.new <- gsub(pattern, "", sample.levels)

# Extract B1, B2, B3 from new vector
batch <- gsub(".*B", "B", sample.levels.new)

Seurat_Parse12sample <- annotate(Seurat_Parse12sample, annotations = batch, to_label = "sample",annotation_name = "Batch")

table(Seurat_Parse12sample$Batch)
table(Seurat_Parse12sample$Batch,Seurat_Parse12sample$sample)

# add the cell line name
# sample vector is still the input vector
# Define regular expression to remove B1, B2, and B3
pattern <- "B[1-3]$"

# Use gsub() to remove B1, B2, and B3 from original vector
sample.levels.new <- gsub(pattern, "", sample.levels)

# Extract starting values from new vector
ipscline <- gsub("B[1-3]$", "", sample.levels.new)
ipscline

Seurat_Parse12sample <- annotate(Seurat_Parse12sample, annotations = ipscline, to_label = "sample",annotation_name = "IPSC_Line")

table(Seurat_Parse12sample$IPSC_Line)
table(Seurat_Parse12sample$IPSC_Line,Seurat_Parse12sample$sample)

# add disease status
# we need to know the order of the lines

Idents(Seurat_Parse12sample) <- "IPSC_Line"
line.levels <- levels(Seurat_Parse12sample)
line.levels

PDstatus <- c("PD","PD","Con","Con")  # if TD07 and 2965 are PD lines and TD22 and 3448 are control lines
Seurat_Parse12sample <- annotate(Seurat_Parse12sample, annotations = PDstatus, to_label = "IPSC_Line",annotation_name = "DiseaseStatus")

table(Seurat_Parse12sample$DiseaseStatus)

 
table(Seurat_Parse12sample$DiseaseStatus,Seurat_Parse12sample$IPSC_Line)

table(Seurat_Parse12sample$Batch,Seurat_Parse12sample$IPSC_Line)




```


Save info 
```{r}
saveRDS(Seurat_Parse12sample, "Parse12sample4lines3batchJuly7.RDS")

seu <- Seurat_Parse12sample
rm(Seurat_Parse12sample)

# plots of in a different workbook for the pre aligned data

```

Align the cell lines and batches, we will align across the 12 samples

```{r}
# make a list of seurat objects by our cell type variable
sublist <- SplitObject(seu, split.by = "sample")
# normalize and find variable features
for (i in 1:length(sublist)){
  sublist[[i]] <- NormalizeData(sublist[[i]], verbose = FALSE)
  sublist[[i]] <- FindVariableFeatures(sublist[[i]], selection.method = "vst")
}
# Create an empty Seurat object to store the integrated data
# Take the first Seurat object from the list as the starting point
integrated_seurat <- subset(sublist[[1]])


# Iterate over the list of Seurat objects
for (i in 1:length(sublist)) {
  # Rename the 'orig.ident' metadata inside the seurat object to match the object name in the list
  sublist[[i]]$orig.ident <- names(sublist)[i]

}

sample.list <- sublist
for (i in 1:length(sample.list)) {
  # Normalize and scale the data
  sample.list[[i]] <- NormalizeData(sample.list[[i]], verbose = FALSE)
  sample.list[[i]] <- ScaleData(sample.list[[i]], verbose = FALSE)
  # Find variable features
  sample.list[[i]] <- FindVariableFeatures(sample.list[[i]], selection.method = "vst")
  # Get the variable features
  variable_features <- VariableFeatures(sample.list[[i]])
  # Run PCA with the variable features
  sample.list[[i]] <- RunPCA(sample.list[[i]], verbose = FALSE, npcs = 30, features = variable_features)
}

int.anchors <- FindIntegrationAnchors(object.list = sample.list, dims = 1:30, reduction = "rpca")
integrated_seurat <- IntegrateData(anchorset = int.anchors,  dims = 1:30)
# 
# must set the k weight to the lowest cell count 
# in the parse sample we have over 1530 cells in the smallest count so we don't have to change the k from the 100 default




```


Now we need to run the workflow on the integrated object

```{r}
DefaultAssay(integrated_seurat) <- "integrated"
integrated_seurat <- ScaleData(integrated_seurat, verbose = FALSE)
# only the integrated features will be the pca input

integrated_seurat <- RunPCA(integrated_seurat, npcs = 20, verbose = FALSE)
integrated_seurat <- RunUMAP(integrated_seurat, reduction = "pca", dims = 1:20, n.neighbors = 81)

```

Have a look at the new UMAP

```{r}
DimPlot(integrated_seurat, group.by = 'sample')
DimPlot(integrated_seurat, group.by = 'Batch')
DimPlot(integrated_seurat, group.by = 'DiseaseStatus')
DimPlot(integrated_seurat, group.by = 'IPSC_Line')



```

```{r}
saveRDS(integrated_seurat, "/Users/rhalenathomas/Documents/Data/scRNAseq/ParseExample/Experiment1-mini12/Integrated12samples.RDS")
# setwd("~/Documents/Data/scRNAseq/ParseExample/Experiment1-mini12")
integrated_seurat <- readRDS("/Users/rhalenathomas/Documents/Data/scRNAseq/ParseExample/Experiment1-mini12/Integrated12samples.RDS")

```

Find new clusters

```{r}
DefaultAssay(integrated_seurat) <- "integrated"
integrated_seurat <- FindNeighbors(integrated_seurat, dims = 1:20, k.param = 81)
integrated_seurat <- FindClusters(integrated_seurat, resolution = c(0,0.3,0.6,1) )

```

```{r}
library(clustree)
clustree(integrated_seurat)


```

```{r}

DimPlot(integrated_seurat, group.by = "integrated_snn_res.0.3")
DimPlot(integrated_seurat, group.by = "integrated_snn_res.0.6")

```
```{r}
DimPlot(integrated_seurat, group.by = "integrated_snn_res.0.3", split.by = "DiseaseStatus")

```
```{r}
table(integrated_seurat$DiseaseStatus)

```




Annotate clusters res 0.3

```{r}

Idents(integrated_seurat) <- "integrated_snn_res.0.3"
ClusterMarkers <- FindAllMarkers(integrated_seurat, only.pos = TRUE)
write.csv(ClusterMarkers, "clusterMarkersres03Integrated.csv")

```



```{r}
library(enrichR)
setEnrichrSite("Enrichr") # Human genes
# list of all the databases
# get the possible libraries
dbs <- listEnrichrDbs()

# this will list the possible libraries
dbs

# select libraries with cell types
db <- c('CellMarker_Augmented_2021','Azimuth_Cell_Types_2021')

# function for a quick look
checkCelltypes <- function(cluster_num = 0){
  clusterX <- ClusterMarkers %>% filter(cluster == cluster_num & avg_log2FC > 0.25)
  genes <- clusterX$gene
  # the cell type libraries
  # get the results for each library
  clusterX.cell <- enrichr(genes, databases = db)
  # visualize the results
print(plotEnrich(clusterX.cell[[1]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value", title = 'CellMarker_Augmented_2021'))
print(plotEnrich(clusterX.cell[[2]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value", title = 'Azimuth_Cell_Types_2021'))

}

```
```{r}
#heatmap of top markers
top3 <- ClusterMarkers %>% group_by(cluster) %>% top_n(n=3, wt =avg_log2FC)
DoHeatmap(integrated_seurat, features = top3$gene, size = 3, angle = 90, group.by = "integrated_snn_res.0.3")


```

```{r}
table(ClusterMarkers$cluster)
```




Check each cluster quickly

```{r}
checkCelltypes(cluster_num = 3)

```

Look at some expression lists

```{r}

da_neurons <- c("TH","SLC6A3","SLC18A2","SOX6","NDNF","SNCG","ALDH1A1","CALB1","TACR2","SLC17A6","SLC32A1","OTX2","GRP","LPL","CCK","VIP")
NPC_orStemLike <- c("DCX","NEUROD1","TBR1","PCNA","MKI67","SOX2","NES","PAX6","MASH1")
mature_neurons = c("RBFOX3","SYP","DLG45","VAMP1","VAMP2","TUBB3","SYT1","BSN","HOMER1","SLC17A6")
excitatory_neurons = c("GRIA2","GRIA1","GRIA4","GRIN1","GRIN2B","GRIN2A","GRIN3A","GRIN3","GRIP1","CAMK2A")
inhbitory_neurons = inh = c("GAD1","GAD2", "GAT1","PVALB","GABR2","GABR1","GBRR1","GABRB2","GABRB1","GABRB3","GABRA6","GABRA1","GABRA4","TRAK2")
astrocytes <- c("GFAP","S100B","AQP4","APOE", "SOX9","SLC1A3")
oligodendrocytes <- c("MBP","MOG","OLIG1","OLIG2","SOX10")
opc <- 
radial_glia <- c("PTPRC","AIF1","ADGRE1", "VIM", "TNC","PTPRZ1","FAM107A","HOPX","LIFR",
              "ITGB5","IL6ST","SLC1A3")
epithelial <- c("HES1","HES5","SOX2","SOX10","NES","CDH1","NOTCH1")

microglia <- c("IBA1","P2RY12","P2RY13","TREM119", "GPR34","SIGLECH","TREM2",
               "CX3CR1","FCRLS","OLFML3","HEXB","TGFBR1", "SALL1","MERTK",
               "PROS1")

features_list <- c("MKI67","SOX2","POU5F1","DLX2","PAX6","SOX9","HES1","NES","RBFOX3","MAP2","NCAM1","CD24","GRIA2","GRIN2B","GABBR1","GAD1","GAD2","GABRA1","GABRB2","TH","ALDH1A1","LMX1B","NR4A2","CORIN","CALB1","KCNJ6","CXCR4","ITGA6","SLC1A3","CD44","AQP4","S100B", "PDGFRA","OLIG2","MBP","CLDN11","VIM","VCAM1")

short_list <- c("MKI67","SOX9","HES1","NES","DLX2","RBFOX3","MAP2","TH","CALB1","KCNJ6","SLC1A3","CD44","AQP4","S100B","OLIG2","MBP","VIM")

```


```{r}
Idents(integrated_seurat) <- "integrated_snn_res.0.3"

for (i in da_neurons) {
  print(FeaturePlot(integrated_seurat, features = i, min.cutoff = 'q1', max.cutoff = 'q97', label = TRUE))
}

```

```{r}
Idents(integrated_seurat) <- "integrated_snn_res.0.3"

for (i in NPC_orStemLike) {
  print(FeaturePlot(integrated_seurat, features = i, min.cutoff = 'q1', max.cutoff = 'q97', label = TRUE))
}
```

```{r}
Idents(integrated_seurat) <- "integrated_snn_res.0.3"

for (i in astrocytes) {
  print(FeaturePlot(integrated_seurat, features = i, min.cutoff = 'q1', max.cutoff = 'q97', label = TRUE))
}
```

```{r}
Idents(integrated_seurat) <- "integrated_snn_res.0.3"

for (i in radial_glia) {
  print(FeaturePlot(integrated_seurat, features = i, min.cutoff = 'q1', max.cutoff = 'q97', label = TRUE))
}

```


```{r}
Idents(integrated_seurat) <- "integrated_snn_res.0.3"

for (i in mature_neurons) {
  print(FeaturePlot(integrated_seurat, features = i, min.cutoff = 'q1', max.cutoff = 'q97', label = TRUE))
}


```

```{r}

Idents(integrated_seurat) <- "integrated_snn_res.0.3"

for (i in excitatory_neurons) {
  print(FeaturePlot(integrated_seurat, features = i, min.cutoff = 'q1', max.cutoff = 'q97', label = TRUE))
}


```
Add annotations - first pass
NPC-stem
NPC-glia
NPC-SOX6
Neurons-Glut
Progenitors-div
NPC-SOX2-OXT-fibro
Neural-Stem
stem cell
Neuron-GABA
Neuron-epithelial

```{r}
celltypes1 <- c("NPC-stem","NPC-glia","NPC-SOX6","Neurons-Glut","Progenitors-div",
                "NPC-fibro","Neural-Stem","Stem","Neurons-GABA","Neural-epi")  
integrated_seurat <- annotate(integrated_seurat, annotations = celltypes1, to_label = "integrated_snn_res.0.3",annotation_name = "Celltypes1")

DimPlot(integrated_seurat, label = TRUE)

```

```{r}

celltypes2 <- c("NPC","NPC","NPC","Neurons","NPC",
                "NPC","NPC","Stem","Neurons","Epithelial")  

integrated_seurat <- annotate(integrated_seurat, annotations = celltypes2, to_label = "integrated_snn_res.0.3",annotation_name = "Celltypes2")

DimPlot(integrated_seurat, label = TRUE)

```

```{r}


DimPlot(integrated_seurat, split.by = "DiseaseStatus")
DimPlot(integrated_seurat, split.by = "DiseaseStatus", group.by = "Celltypes1")


```

```{r}
celltypes3 <- c("NPC","NPC","NPC","Neurons","NPC-div",
                "Neuro-NPC","Neural-Stem","Stem","Neurons","Neural-epi")  
integrated_seurat <- annotate(integrated_seurat, annotations = celltypes3, to_label = "integrated_snn_res.0.3",annotation_name = "Celltypes3")

DimPlot(integrated_seurat, label = TRUE)
```
```{r}
table(integrated_seurat$IPSC_Line)
```



DEG in cell types 3 groups

```{r}
DefaultAssay(integrated_seurat) <- "RNA"
Idents(integrated_seurat) <- "Celltypes3"
levels(integrated_seurat)
seu_sub <- subset(integrated_seurat, idents = "Neurons")
dim(seu_sub)
table(seu_sub$IPSC_Line)
seu_sub <- ScaleData(seu_sub)
seu_sub <- NormalizeData(seu_sub)

Idents(seu_sub) <- "DiseaseStatus"
levels(seu_sub)
DGE <- FindMarkers(seu_sub, ident.1 = "PD", ident.2 = "Con")
DimPlot(seu_sub, group.by = "DiseaseStatus")


```
```{r}
table(seu_sub$DiseaseStatus)
```


Neurons DGE volcano plot

```{r, fig.height= 3}

library(EnhancedVolcano)
EnhancedVolcano(DGE,
    lab = rownames(DGE),
    #xlim = c(-0.25,0.25),
    x = 'avg_log2FC',
    y = 'p_val_adj',
    pCutoff = 0.000001,
    FCcutoff = 1,
    pointSize = 3.0,
    labSize = 6.0)



```

```{r}
ft <- c("MGAT4C","TBR1","SLC4A10","RELN","PTPRD",
        "GPC5","SIM1","ONECUT2","EBF1","LMX1A")

DotPlot(seu_sub, features = ft) +RotatedAxis()

head(DGE)
# filter for significant pvalues
DGE.ft <- DGE %>% filter(p_val_adj <= 0.01)
up <- DGE.ft %>% filter(avg_log2FC > 0.1)
down <- DGE.ft %>% filter(avg_log2FC < -0.1)


```

DGE in NPCs: NPC-div

```{r}
DefaultAssay(integrated_seurat) <- "RNA"
Idents(integrated_seurat) <- "Celltypes3"
levels(integrated_seurat)
seu_sub <- subset(integrated_seurat, idents = "NPC-div")
dim(seu_sub)
table(seu_sub$IPSC_Line)
table(seu_sub$DiseaseStatus)
seu_sub <- ScaleData(seu_sub)
seu_sub <- NormalizeData(seu_sub)

Idents(seu_sub) <- "DiseaseStatus"
levels(seu_sub)
DGE <- FindMarkers(seu_sub, ident.1 = "PD", ident.2 = "Con")
DimPlot(seu_sub, group.by = "DiseaseStatus")


```

```{r, fig.height=3}
EnhancedVolcano(DGE,
    lab = rownames(DGE),
    #xlim = c(-0.25,0.25),
    x = 'avg_log2FC',
    y = 'p_val_adj',
    pCutoff = 0.000001,
    FCcutoff = 1,
    pointSize = 3.0,
    labSize = 6.0)
```
```{r}
ft <- c("TEME132C","TMEM132D","GPC3","NRG1","PDE1A",
        "TTC6","SPON1","NTN1","PK1B","BNC2")

DotPlot(seu_sub, features = ft) +RotatedAxis()

#head(DGE)
# filter for significant pvalues
DGE.ft <- DGE %>% filter(p_val_adj <= 0.01)
dim(DGE.ft)
up <- DGE.ft %>% filter(avg_log2FC > 0.1)
dim(up)
down <- DGE.ft %>% filter(avg_log2FC < -0.1)
dim(down)
```

```{r}

# main NPC group
DefaultAssay(integrated_seurat) <- "RNA"
Idents(integrated_seurat) <- "Celltypes3"
levels(integrated_seurat)
seu_sub <- subset(integrated_seurat, idents = "NPC")
dim(seu_sub)
table(seu_sub$IPSC_Line)
table(seu_sub$DiseaseStatus)
#seu_sub <- ScaleData(seu_sub)
seu_sub <- NormalizeData(seu_sub)

Idents(seu_sub) <- "DiseaseStatus"
levels(seu_sub)
DGE <- FindMarkers(seu_sub, ident.1 = "PD", ident.2 = "Con")
DimPlot(seu_sub, group.by = "DiseaseStatus")

```
```{r, fig.height=3}
library(EnhancedVolcano)
EnhancedVolcano(DGE,
    lab = rownames(DGE),
    xlim = c(-3,3),
    x = 'avg_log2FC',
    y = 'p_val_adj',
    pCutoff = 0.01,
    FCcutoff = 1,
    pointSize = 3.0,
    labSize = 6.0)

```

```{r}

ft <- c("RBFOX1","TMEM132C","GPC3","NRG1","LINGO2",
        "TTC6","SLIT1","CHN2","NTN1","ALK")

DotPlot(seu_sub, features = ft) +RotatedAxis()

#head(DGE)
# filter for significant pvalues
DGE.ft <- DGE %>% filter(p_val_adj <= 0.01)
dim(DGE.ft)
up <- DGE.ft %>% filter(avg_log2FC > 0.1)
dim(up)
down <- DGE.ft %>% filter(avg_log2FC < -0.1)
dim(down)
```

From the pseudo bulk
```{r}
ft <- c("CLRN1","NR2E1","IFI44","DMRT3","FEZF2",
        "KCNJ16","RGPD2","CORIN","GALR1","SIM1")

DotPlot(seu_sub, features = ft) +RotatedAxis()
```


```{r}

library(enrichR)
setEnrichrSite("Enrichr") # Human genes
# list of all the databases
# get the possible libraries
dbs <- listEnrichrDbs()

# this will list the possible libraries
dbs

# select libraries with cell types
db <- c('KEGG_2019_Human','GWAS_Catalog_2019',"GO_Biological_Process_2023",
        "GO_Cellular_Component_2023","GO_Molecular_Function_2023")

# use only the significant after p adjusted genes
# add genes column to DGE it is X if saved to csv
DGE.ft$X <- rownames(DGE.ft)

# function for a quick look
getGSA <- function(dataframe, up_or_down = "up", LFCthresh = 0.01,
                          pval_thresh = 0.05){
  if(up_or_down == "up"){
      genelist <- dataframe %>% filter(avg_log2FC >= LFCthresh & p_val_adj < pval_thresh)
  }else if (up_or_down == "down") {
      genelist <- dataframe %>% filter(avg_log2FC <= LFCthresh & p_val_adj < pval_thresh)
  } else {
    genelist <- dataframe %>% filter(p_val_adj < pval_thresh)
  }
genes <- genelist$X
  # the cell type libraries
  # get the results for each library
  results <- enrichr(genes, databases = db)
  # visualize the results
  
print(plotEnrich(results[[1]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value", title = 'KEGG'))
print(plotEnrich(results[[2]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value", title = 'GWAS'))
print(plotEnrich(results[[3]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value", title = 'GObio'))
print(plotEnrich(results[[4]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value", title = 'GOcell'))
print(plotEnrich(results[[5]], showTerms = 20, numChar = 40, y = "Count", orderBy = "P.value", title = 'GOmol'))
return(results)
}

DGEup <- getGSA(DGE.ft, up_or_down = "up",
                LFCthresh = 0.02,
                pval_thresh = 0.05)


```

```{r}
DGEdown <- getGSA(DGE.ft, up_or_down = "down",
                LFCthresh = -0.02,
                pval_thresh = 0.05)
```

```{r}
DGEboth <- getGSA(DGE.ft, up_or_down = "both",
                LFCthresh = 0,
                pval_thresh = 0.01)

```

The DGE.ft to compare to with the pseudobulk results

```{r}
# write the DGE
write.csv(DGE.ft, "/Users/rhalenathomas/Documents/Data/scRNAseq/ParseExample/Experiment1-mini12/DGEseuratNPCs.csv")


```



Create a sum of counts data table 

```{r}
sum_counts <- AggregateExpression(integrated_seurat, assay = "RNA", group.by = "sample", add.ident = "sample")
# this seems to group by the group and the active ident together

colnames(integrated_seurat@meta.data)
dim(sum_counts$RNA)
sum_counts_df <- as.data.frame(sum_counts$RNA)
dim(sum_counts_df)
write.csv(sum_counts_df, "/Users/rhalenathomas/Documents/Data/scRNAseq/ParseExample/Experiment1-mini12/SumCounts_Integrated_by_sample_celltype.csv")
sum_counts_df[1:4,1:5]



```


Try to separately variables in the aggregation

```{r}
colnames(integrated_seurat@meta.data)

sum_counts <- AggregateExpression(integrated_seurat, assay = "RNA", group.by = c("Batch","IPSC_Line","DiseaseStatus"))
# this seems to group by the group and the active ident together



colnames(integrated_seurat@meta.data)
dim(sum_counts$RNA)
sum_counts_df <- as.data.frame(sum_counts$RNA)
dim(sum_counts_df)

# all cell types are integrated

write.csv(sum_counts_df, "/Users/rhalenathomas/Documents/Data/scRNAseq/ParseExample/Experiment1-mini12/SumCounts_Integrated_allCelltypes_12samples.csv")


```


Try to make the MDS plots and the DESeq2 DGE

```{r}
library( "DESeq2" )
library(ggplot2)

```

Visualize groups

```{r}
DimPlot(integrated_seurat, group.by = "Celltypes3", split.by = "Batch")
DimPlot(integrated_seurat, group.by = "Celltypes3", split.by = "DiseaseStatus")
DimPlot(integrated_seurat, group.by = "Celltypes3", 
        split.by = "sample", ncol = 4)

DimPlot(seu_sub, group.by = "DiseaseStatus",
        split.by = "Batch")
DimPlot(seu_sub, 
        split.by = "sample", ncol = 4)

```



